src/operator/distinct.rs

```diff
@@ -0,0 +1,288 @@
+                )
+            })
+            .clone()
+    }
+
+    /// Incremental nested version of the [`Distinct`] operator.
+    pub fn distinct_incremental_nested<V, W>(&self) -> Stream<Circuit<P>, <Z as SharedRef>::Target>
+    where
+        V: KeyProperties,
+        W: ZRingValue,
+        Z: SharedRef + 'static,
+        <Z as SharedRef>::Target: ZSet<V, W> + SharedRef<Target = Z::Target>,
+        for<'a> &'a <Z as SharedRef>::Target: IntoIterator<Item = (&'a V, &'a W)>,
+    {
+        self.integrate_nested()
+            .distinct_incremental()
+            .differentiate_nested()
+    }
+}
+
+/// Distinct operator changes all weights in the support of a Z-set to 1.
+pub struct Distinct<V, W, Z> {
+    _type: PhantomData<(V, W, Z)>,
+}
+
+impl<V, W, Z> Distinct<V, W, Z> {
+    pub fn new() -> Self {
+        Self { _type: PhantomData }
+    }
+}
+
+impl<V, W, Z> Default for Distinct<V, W, Z> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<V, W, Z> Operator for Distinct<V, W, Z>
+where
+    V: 'static,
+    W: 'static,
+    Z: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Distinct")
+    }
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<V, W, Z> UnaryOperator<Z, Z> for Distinct<V, W, Z>
+where
+    V: KeyProperties,
+    W: ZRingValue,
+    Z: ZSet<V, W> + 'static,
+{
+    fn eval(&mut self, i: &Z) -> Z {
+        i.distinct()
+    }
+
+    fn eval_owned(&mut self, i: Z) -> Z {
+        i.distinct_owned()
+    }
+}
+
+/// Incremental version of the distinct operator.
+///
+/// Takes a stream `a` of changes to relation `A` and a stream with delayed
+/// value of `A`: `z^-1(A) = a.integrate().delay()` and computes
+/// `distinct(A) - distinct(z^-1(A))` incrementally, by only considering
+/// values in the support of `a`.
+struct DistinctIncremental<V, W, Z> {
+    _type: PhantomData<(V, W, Z)>,
+}
+
+impl<V, W, Z> DistinctIncremental<V, W, Z> {
+    pub fn new() -> Self {
+        Self { _type: PhantomData }
+    }
+}
+
+impl<V, W, Z> Default for DistinctIncremental<V, W, Z> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<V, W, Z> Operator for DistinctIncremental<V, W, Z>
+where
+    V: 'static,
+    W: 'static,
+    Z: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("DistinctIncremental")
+    }
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<V, W, Z> BinaryOperator<Z, Z, Z> for DistinctIncremental<V, W, Z>
+where
+    V: KeyProperties,
+    W: ZRingValue,
+    for<'a> &'a Z: IntoIterator<Item = (&'a V, &'a W)>,
+    Z: ZSet<V, W> + 'static,
+{
+    fn eval(&mut self, delta: &Z, delayed_integral: &Z) -> Z {
+        let mut result = Z::empty();
+
+        for (v, w) in delta.into_iter() {
+            let weight = delayed_integral.lookup(v);
+
+            if weight.is_zero() {
```

Comment:
I don't think this is general enough.
This works if the input is a set, but once you move distincts around inputs may be multisets.
So it is safer to use the formulas in the paper, with <= 0 and >= 0.
I realize that ZRingValue does not have le0, could be a bit more efficient than negating.

Reply:
Sorry, I don't understand, why won't this work for multisets?

Reply:
The paper says that the weight is 1 when it changes from negative to positive. You require it to be exactly 0.

Reply:
I guess because we want distinct to map negative values to 0. Fair enough.  Is that an arbitrary choice or is there a reason theory wouldn't work if distinct(-1) was was 1 instead of 0.

Reply:
It won't work because it does not compute set difference correctly.

Reply:
Should be fixed now.